src/contracts/MslV5.ts

```diff
@@ -310,17 +310,61 @@ export class MslV5 extends BaseContract<typeof multiSourceLoanABIV5> {
     return lockupTimeSeconds - ellapsedSeconds;
   }
 
-  async refinanceFullLoan({
-    offer,
-    signature,
-    loan,
+  async refinanceBatch({
+    renegotiationId,
+    refinancings,
   }: {
-    offer: RenegotiationV5;
-    signature: Hash;
-    loan: LoanV5;
+    renegotiationId: bigint;
+    refinancings: {
+      loan: LoanV5;
+      newAprBps: bigint;
+      sources: {
+        source: LoanV5['source'][number];
+        refinancingPrincipal: bigint;
+      }[];
+    }[];
   }) {
-    const txHash = await this.safeContractWrite.refinanceFull([offer, loan, signature]);
+    // Generate multicall encoded function data for (renegotiation offer, loan) pairs
+    const data = refinancings.map(({ loan, sources, newAprBps }, index) => {
+      const targetPrincipal = loan.source.map(({ principalAmount, loanId }) => {
+        const refinancingSource = sources.find(({ source }) => source.loanId === loanId);
+        return refinancingSource
+          ? principalAmount - refinancingSource.refinancingPrincipal
+          : principalAmount;
+      });
+      const refinancingPrincipalAmount = sources.reduce(
+        (acc, { refinancingPrincipal }) => acc + refinancingPrincipal,
+        0n,
+      );
+
+      const offer = {
+        renegotiationId: renegotiationId + BigInt(index),
+        loanId: loan.source[0].loanId,
+        lender: this.wallet.account.address,
+        fee: 0n,
+        targetPrincipal,
+        principalAmount: refinancingPrincipalAmount,
+        aprBps: newAprBps,
+        expirationTime: 0n,
+        duration: 0n,
+      };
```

Comment:
this is repeated

Reply:
offer generation is the same in v4 and v5, the only thing different is the multicall.
I would probably abstract one of the two

Reply:
Let's do it in another PR.